### PR TITLE
Logic for processing VIIRS snow data

### PR DESCRIPTION
**Pull request by dgboss**
The general flow is:

1. Check the api database for the date of the last successful processing of VIIRS snow data.
2. Download bc_boundary.geojson from S3 into a temp directory
3. Download 'granules' (tif files representing satellite passes over BC) into another temp directory
4. Create a single image (aka mosaic) from the downloaded tifs
5. Clip the single image using the bc_boundary.geojson file
6. Upload the clipped file to S3 storage

Occasionally there is a delay in availability of the data. If we encounter this we exit gracefully and try again the next time the cron job runs. This is a pretty specific HTTP status code (501). All other errors get logged and a notification sent to RocketChat.

You can view the output files in S3 at /snow_coverage/{YYYY-mm-dd}/*.tif. The most recent folder should ideally correspond with the most recent available folder at https://n5eil01u.ecs.nsidc.org/VIIRS/VJ110A1.002/.
# Test Links:
[Landing Page](https://wps-pr-3332.apps.silver.devops.gov.bc.ca/)
[MoreCast 2.0](https://wps-pr-3332.apps.silver.devops.gov.bc.ca/morecast-2)
[Percentile Calculator](https://wps-pr-3332.apps.silver.devops.gov.bc.ca/percentile-calculator)
[MoreCast](https://wps-pr-3332.apps.silver.devops.gov.bc.ca/morecast)
[C-Haines](https://wps-pr-3332.apps.silver.devops.gov.bc.ca/c-haines)
[FireBat](https://wps-pr-3332.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator)
[FireBat bookmark](https://wps-pr-3332.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator?s=266&f=c5&c=NaN&w=20,s=286&f=c7&c=NaN&w=16,s=1055&f=c7&c=NaN&w=NaN,s=305&f=c7&c=NaN&w=NaN,s=344&f=c5&c=NaN&w=NaN,s=346&f=c7&c=NaN&w=NaN,s=328&f=c7&c=NaN&w=NaN,s=1399&f=c7&c=NaN&w=NaN,s=334&f=c7&c=NaN&w=NaN,s=1082&f=c3&c=NaN&w=NaN,s=388&f=c7&c=NaN&w=NaN,s=309&f=c7&c=NaN&w=16,s=306&f=c7&c=NaN&w=NaN,s=1029&f=c7&c=NaN&w=NaN,s=298&f=c7&c=NaN&w=NaN,s=836&f=c7&c=NaN&w=NaN,s=9999&f=c7&c=NaN&w=NaN)
[Auto Spatial Advisory (ASA)](https://wps-pr-3332.apps.silver.devops.gov.bc.ca/auto-spatial-advisory)
[HFI Calculator](https://wps-pr-3332.apps.silver.devops.gov.bc.ca/hfi-calculator)

### CODE REVIEW

File: api/app/db/models/weather_models.py [hunk 1]
```diff
@@ -184,7 +184,7 @@ class ModelRunPrediction(Base):
     # The date and time to which the prediction applies.
     prediction_timestamp = Column(TZTimeStamp, nullable=False, index=True)
     # The station code representing the location (aka weather station).
-    station_code = Column(Integer, nullable=True)
+    station_code = Column(Integer, nullable=False)
```
Comment (conbrad) [inline api/app/db/models/weather_models.py:187]: I'm guessing this should have always been `False`, correct?

File: api/app/jobs/viirs_snow.py [hunk 2]
```diff
@@ -1,11 +1,222 @@
...
+        """
+        # Interesting flow required by the NSIDC API. We need to issue a request without credentials which will fail
+        # with a 401. The URL in the response object is the authentication endpoint, so we issue a request to it with basic auth.
+        # Now we are authenticated with a session cookie and can make additional requests to the API.
+        # Furthermore, it looks like we're stuck with synchronous requests.
+        logger.info(f"Downloading VIIRS snow coverage data for date: {for_date.strftime('%Y-%m-%d')}")
+        session = requests.session()
+        resp = session.get(f"https://n5eil02u.ecs.nsidc.org/egi/capabilities/{SHORT_NAME}.{PRODUCT_VERSION}.xml")
+        session.get(resp.url, auth=(config.get("NASA_EARTHDATA_USER"),config.get("NASA_EARTHDATA_PWD")))
+        # Check if request was successful
```
Comment (conbrad) [inline api/app/jobs/viirs_snow.py:62]: Looks like we do 2 requests here? One to authenticate, then one to get the data. Where are we checking that the auth request was successful?

File: api/app/jobs/viirs_snow.py [hunk 3]
```diff
@@ -1,11 +1,222 @@
...
+        """
+        async with get_client() as (client, bucket):
+            bc_boundary = await client.get_object(Bucket=bucket, Key="bc_boundary/bc_boundary.geojson")
+            data = await bc_boundary['Body'].read()
+            bc_boundary_geojson_path = os.path.join(path, "bc_boundary.geojson")
+            with open(bc_boundary_geojson_path, "wb") as file:
+                file.write(data)
+
+
+    async def _save_clipped_snow_coverage_moasic_to_s3(self, for_date: date, path: str):
```
Comment (conbrad) [inline api/app/jobs/viirs_snow.py:124]:
```suggestion
    async def _save_clipped_snow_coverage_mosaic_to_s3(self, for_date: date, path: str):
```

File: api/app/jobs/viirs_snow.py [hunk 4]
```diff
@@ -1,11 +1,222 @@
...
+                        break
+                    else:
+                        # If a different HTTPError occurred re-raise and let the next exception handler send a notification to RocketChat.
+                        raise http_error
+                next_date = next_date + timedelta(days=1)
+
+
 def main():
-    logger.info("***************Cron job started**************")
+    """ Kicks off asyncronous processing of VIIRS snow coverage data.
```
Comment (conbrad) [inline api/app/jobs/viirs_snow.py:200]:
```suggestion
    """ Kicks off asynchronous processing of VIIRS snow coverage data.
```

**Review by conbrad — APPROVED** [review]

File: api/app/db/models/weather_models.py (hunk 1 above)
Comment (dgboss) [reply api/app/db/models/weather_models.py:187]: Yep. I think I originally had it set to True as a column was being added to an existing table so there was the possibility of null values and providing a default didn't make sense. In the end I deleted the contents of the table in the migration and had the migration set nullable=False. This change brings the code into alignment with the current schema of the db.

File: api/app/jobs/viirs_snow.py (hunk 2 above)
Comment (dgboss) [reply api/app/jobs/viirs_snow.py:62]: Oops, that's a leftover comment at this spot that I'll remove. The check to see if we've made a successful request actually occurs below with `request.raise_for_status()`. If the request for data returns anything but a 200 this call with raise an error which we subsequently handle.

Comment (dgboss) [inline api/app/jobs/viirs_snow.py:62]:
```suggestion
```

**Pull request merged**